File: runtime/java/src/main/java/io/nightseam/runtime/WireFrames.java
package io.nightseam.runtime;

import dev.bitspark.bitwire.*;

import dev.bitspark.bitwire.*;
import java.util.LinkedHashMap;
import java.util.Map;

/** Conversion at the Nightseam JSON profile boundary. Wire composition needs no conversion. */
public final class WireFrames {
    private WireFrames() {}
    public static Map<String,Object> fields(ProfileFrame frame) {
        var out = new LinkedHashMap<String,Object>();
        out.put("version", frame.version());
        out.put("kind", frame.kind().name().toLowerCase(java.util.Locale.ROOT));
        if (frame.traceparent() != null) out.put("traceparent", frame.traceparent());
        if (frame.tracestate() != null) out.put("tracestate", frame.tracestate());
        switch (frame) {
            case ProfileFrame.Request request -> {
                out.put("id", request.id()); out.put("params", Json.parse(request.params().json()));
                if (request.meta() != null) out.put("meta", request.meta());
            }
            case ProfileFrame.Event event -> {
                out.put("data", Json.parse(event.data().json()));
                if (event.meta() != null) out.put("meta", event.meta());
            }
            case ProfileFrame.Cancel cancel -> out.put("id", cancel.id());
            case ProfileFrame.Response response -> {
                out.put("id", response.id());
                if (response.result() != null) out.put("result", Json.parse(response.result().json()));
                else {
                    var error = new LinkedHashMap<String,Object>();
                    error.put("code", response.error().code()); error.put("message", response.error().message());
                    if (response.error().data() != null) error.put("data", Json.parse(response.error().data().json()));
                    out.put("error", error);
                }
            }
        }
        return out;
    }
    @SuppressWarnings("unchecked")
    public static ProfileFrame frame(Map<String,Object> fields) {
        String id = (String) fields.get("id"), parent = (String) fields.get("traceparent"), state = (String) fields.get("tracestate");
        Map<String,String> meta = (Map<String,String>) fields.get("meta");
        return switch ((String) fields.get("kind")) {
            case "request" -> new ProfileFrame.Request(id, payload(fields, "params"), meta, parent, state);
            case "event" -> new ProfileFrame.Event(payload(fields, "data"), meta, parent, state);
            case "cancel" -> new ProfileFrame.Cancel(id, parent, state);
            case "response" -> {
                ProfileError error = null;
                if (fields.containsKey("error")) {
                    var value = Json.object(fields.get("error"));
                    error = new ProfileError((String)value.get("code"), (String)value.get("message"),
                        value.containsKey("data") ? new JsonValue(Json.stringify(value.get("data"))) : null);
                }
                yield new ProfileFrame.Response(id, fields.containsKey("result") ? payload(fields, "result") : null, error, parent, state);
            }
            default -> throw new IllegalArgumentException("unknown wire frame kind");
        };
    }
    private static JsonValue payload(Map<String,Object> fields, String key) {
        if (!fields.containsKey(key)) throw new IllegalArgumentException("missing " + key);
        return new JsonValue(Json.stringify(fields.get(key)));
    }
}

File: runtime/hs/src/Nightseam/Runtime/WireFrames.hs
{-# LANGUAGE OverloadedStrings #-}
module Nightseam.Runtime.WireFrames (frameValue, parseWireFrame) where

import Bitwire
import Data.Aeson
import Data.Aeson.Types (parseEither, Parser)
import qualified Data.Aeson.KeyMap as KM
import qualified Data.ByteString.Lazy as L
import Data.Text (Text)
import qualified Data.Text as T

-- Serialization is a profile boundary, not part of path composition.
frameValue :: ProfileFrame -> Value
frameValue (ProfileFrame trace body) = object $
  ["version" .= (1 :: Int)] ++ tracing ++ case body of
    Request ident params meta -> ["kind" .= ("request" :: Text), "id" .= ident, "params" .= payload params] ++ metadata meta
    Event value meta -> ["kind" .= ("event" :: Text), "data" .= payload value] ++ metadata meta
    Cancel ident -> ["kind" .= ("cancel" :: Text), "id" .= ident]
    Response ident result -> ["kind" .= ("response" :: Text), "id" .= ident] ++ either
      (\err -> ["error" .= object (["code" .= errorCode err, "message" .= errorMessage err]
        ++ maybe [] (\value -> ["data" .= payload value]) (errorData err))])
      (\value -> ["result" .= payload value]) result
  where
    payload :: JsonPayload -> Value
    payload value = case eitherDecodeStrict' (jsonBytes value) of
      Right parsed -> parsed
      Left problem -> error ("invalid Bitwire JSON payload: " ++ problem)
    metadata = maybe [] (\meta -> ["meta" .= meta])
    tracing = maybe [] (\value -> ["traceparent" .= value]) (traceparent trace)
      ++ maybe [] (\value -> ["tracestate" .= value]) (tracestate trace)

parseWireFrame :: Value -> Either Text ProfileFrame
parseWireFrame = either (Left . T.pack) Right . parseEither (withObject "wire frame" $ \fields -> do
  version <- fields .: "version" :: Parser Int
  if version /= 1 then fail "invalid profile version" else pure ()
  kind <- fields .: "kind" :: Parser Text
  let json key = do
        value <- maybe (fail "missing JSON payload") pure (KM.lookup key fields)
        pure (JsonPayload (L.toStrict (encode value)))
  trace <- Trace <$> fields .:? "traceparent" <*> fields .:? "tracestate"
  body <- case kind of
    "request" -> Request <$> fields .: "id" <*> json "params" <*> fields .:? "meta"
    "event" -> Event <$> json "data" <*> fields .:? "meta"
    "cancel" -> Cancel <$> fields .: "id"
    "response" -> do
      ident <- fields .: "id"
      outcome <- case (KM.lookup "result" fields, KM.lookup "error" fields) of
        (Just value, Nothing) -> pure (Right (JsonPayload (L.toStrict (encode value))))
        (Nothing, Just value) -> Left <$> withObject "profile error" (\err ->
          ProfileError <$> err .: "code" <*> err .: "message"
            <*> pure (JsonPayload . L.toStrict . encode <$> KM.lookup "data" err)) value
        _ -> fail "response requires exactly one result or error"
      pure (Response ident outcome)
    _ -> fail "invalid profile kind"
  pure (ProfileFrame trace body))
